Linux support and more

albong opened the pull request:
## Summary of changes
There are a lot of changes, however the three major changes are all intertwined as necessary to get things running at all, and then to get things running on linux.

### Major changes
- Added linux support.
    - This mainly boils down to having separate versions of `screenshot.py` in platform specific subdirectories.
    - Added some platform requirements for the screenshot depedencies.
    - Updated tutorial, as this only works when running duckstation as an X11 app. Technically the different compositors expose ways of getting other window contents under Wayland, but its a huge pain. Anyone using this can live with/probably already is running with XWayland.
    - Added different sections for `border_size` in the configuration for linux/windows.
- Changed to a src layout and added pyproject.toml, to be more consistent with current python packaging practices and make it easier for other people to make contributions in the future. (Admittedly, "current" python packaging practices seem to change every other day...)
    - Removed unused dependencies.
        - Some of the things in requirements.txt weren't even transitive dependencies of the direct dependencies. If the desire is to lock dependencies, I suggest doing a clean install in a new environment, and then adding the output of pip freeze to pyproject.toml (or using UV or something).
    - Added `run.py` in the root of the project as an entry point for pyinstaller.
    - Added batch/shell scripts for packaging with pyinstaller.
    - Added `xeupiu` command in pyproject.toml, so that if you install in editable mode you can run with just `xeupiu` from your terminal.
    - Updated imports to be explicit.
    - Needed a different icon file format.
- Tk fixes to resolve strange crashes/hangs on exit.
    - Scheduled the processing to run from the Tk event loop, rather than trying to run an infinite loop inside of the Tk event loop. This is presumably why you needed to call `update()` - you were inside an infinite loop and never yielding back to the Tk event loop. Tk documentation strongly recommends against calling `update()`, which all the overlays did, as this apparently spawns another event loop. I added `XeupiuControlPanel._step()` and followed the documentation to have this function re-schedule itself in the event loop with `after()`.
        - Supposedly `after(0, ...)` will necessarily not run your function immediately but do any other tasks in the queue, however I found that nothing rendered when I did that so it always waits 1ms between steps. I think this is not significant.
    - Creating multiple Tk objects, like `create_master()` did, is also discouraged as it causes problems. Removed this.
    - Startup speed is significantly improved for me.

### Minor changes
- Added some type annotations.
- Added .gitignore and .gitattributes files.
- Added `xdialog` dependency for cross platform error messages.
- Added `--screenshot` to `xeupiu` command to make it easier to debug why OCR was failing.
- Added a markdown guide for development, to make it easier for other people to contribute in the future.
- Added a disclaimer about needing to own things to the tutorial.

## Testing
To test:
- Clone the changes to a new directory.
- Follow development guide to install in a local environment (preferably a new one).
- Validate that running with the `xeupiu` command works and the software functions correctly.
- Validate that packaging/running with pyinstaller works and the software functions correctly.
- Test `xeupiu --screenshot`.

## Comments
- I recognize that nobody appreciates some random person showing up and changing everything with their code, which is what I did by changing the project layout. I just wanted to make it easier to develop and run things in a way that's familiar for python developers, so that more people can easily contribute in the future. Selfishly, its also a lot easier to get an LSP server working correctly when you (editable) install the code as a package.
- Originally I added all the image files from the release as they are necessary to run xeupiu. I removed them from this PR as there are so many of them that it was impossible to review the PR. I will create a second PR after this is merged to add those files - if you need binary files to run your software, they should be in the repo.
    - Can probably make all the character files into a single spritesheet and save on disk space/files. Haven't really looked at how character files are loaded to know if that's easy or not.
- I've actually done quite a bit more development already to support having the overlays move if you drag duckstation around, support minimizing them when duckstation is minimized, and to remove the need for explicitly specifying that you're running in fullscreen. I've also been examining the performance and starting to tweak things - one step is like 40ms on my beefy computer, which is ridiculously high for the amount of work actually being done. Replacing `thefuzz` with a trie, adding numba on a few functions, removing Pillow and just using numpy, should speed things up quite a bit based on my profiling. I'll create PRs for these changes later.
- I still have some weird problem with my keyboard inputs being eaten for a few seconds after I close Xeupiu. I suspect this is a different Tk bug but I'm not sure.

albong committed:
Changed project layout, added linux support.

Changed to src layout. Added pyproject.toml. Made `xeupiu` a command
that can be run. Added binary files needed to run (pulled from release).
Removed unneeded dependencies (mostly indirect, some appear totally
unrelated). Added scripts for building with pyinstaller.

Fixed problems with Tk crashing/hanging. This was mostly related to
using `update()`, which is strongly discouraged by the documentation.
Its also bad to create multiple instances of a Tk object. To make things
work, I schedule `App`'s step in the Tk event loop instead of running an
infinite `while True` loop. Not sure if this was a linux specific
problem, but things seem to start much faster now at least, and I don't
get weird errors when I close the program.

Added support for linux; requires running duckstation with x11. Updated
tutorial with some linux details, and with some comments about running
from source.

vgarciasc committed:
fixed Windows OS detection in xeupiu.py
In XeupiuControlPanel, `sys.platform` was expected to evaluate to "windows", when in fact it evaluates to "win32".

vgarciasc committed:
changing package_executable.bat to reflect new project structure
Now it employs the exact same command as `package_executable.sh`, which is appropriate for the current structure. Final powershell command also removed due to being unnecessary for packaging purposes.

vgarciasc commented:
First of all, sorry for the delayed reply: I wanted to review the PR carefully, and only today managed to get the time to do it. 

Second of all, I understand that some people may not appreciate what you've done here (as you said), but from my part, I have nothing *but* appreciation! I recognize the project was a bit messy (the result of working for so long in something that you're not sure anyone else will care about), so changing the layout was really necessary to make it easier (i.e. possible) for other people to contribute and run from source. So I really appreciate what you've done here.

I've tested everything as you've indicated, and managed to successfully run the project in a new local environment, though some things stood out to me. 

## sys.platform

In `XeupiuControlPanel`'s `__init__`, Windows OS detection expected to see `"windows"`, instead of `"win32"`. I made a commit to fix that.

## package_executable.bat

The `package_executable.bat` did not reflect the new project structure. I changed it to use the exact same command as the `.sh`  version for consistency, and also removed the last powershell command since it was unnecessary. See previous commit.

## Tk stuff

About Tk: I took a look at the Tk docs and indeed, calling `update()` yourself is a bad practice! I didn't know that. Your modifications in that regard look great, and startup is faster for me as well. The only issue I see is that all overlays open immediately after the user runs `xeupiu`, rather than waiting for them to press "Save and run". I believe it is best to run startup after the user presses "Save and run" for two reasons:

1. The green textbox shows "Loading" by default, but it is not loading anything -- it is waiting for the user to press "Save and run";
2. The [showcase video](https://www.youtube.com/watch?v=xrq3V93W8Xs) (which some people use as tutorial) is showing a different behavior, and this might cause some confusion.

The simplest approach I can think of would be to keep everything as-is, but run `self.app = App()` inside the `save_and_run()`. Do you see a problem with this approach, or can think of a better one?

## Final thoughts

Everything else looks great to me, and constitutes very welcome changes. I second your future intended changes: I did my best to make the code as efficient as possible, but I'm sure there is a lot of room for improvement. I also recognize that having tons of character files is not very appropriate -- I kept them this way because it facilitated debugging in the beginning, but keeping them in a single spritesheet is for the best.

albong commented:
Sorry to take so long to respond, I've been traveling.

- I had no idea about `update()` either, I learned a lot about how Tk works.
- I think moving `self.app = App()` inside `save_and_run()` makes sense. I'm getting some weirdness where all overlays don't hide when I do this, but if I increase the time in the call `self.root.after(1, self._step)` at the end of `save_and_run` to 10ms then they disappear. This seems like the same reason I put a wait of at least 1ms in `_step()`, the Tk window events are for some reason not getting processed before `_step()` gets called. I found [a comment to a SO answer](https://stackoverflow.com/a/38817470) that suggests the right thing to do is something like `self.root.after(1, self.root.after_idle, self._step)`, but I don't know enough about Tk and its event loop to be confident that this isn't just a hack. I'm sure there's a correct way to do "run this function after the windows refresh", I'll do a little more reading later today and see if I can find something.

albong committed:
Delayed app object construction until the user clicks "save and run".

Changed how the step funciton is scheduled in the Tk event loop. Used a
combination of after_idle() with after() so that idle events get
processed - mainly the issue was that the alpha setting for the overlays
didn't have time to update so some of the overlays were hanging around.

albong commented:
Okay, I think this should be working now and be ready to merge.

- Moved `self.app = App()` inside of `save_and_run()`
- Changed `self.root.after(...)` to `self.root.after_idle(self.after, ...)`. This is apparently the correct way to do "finish rendering, then do my work". What was happening is: changes to attributes aren't guaranteed to happen immediately, and the changes to the alpha weren't getting processed because `_step()` was constantly rescheduling itself too fast for the attribute change events to get processed. When `App` was constructed earlier, the delay between the overlay creation and the user clicking "save and run" was enough time for the events to process.
- Fixed bug with the sleep time, I'm an idiot and I converted nanoseconds to milliseconds wrong.
- Added "Loading..." to the text log, because there's a slight delay after clicking "save and run" due to loading the files

vgarciasc commented:
Looks great to me! Now that you've explained it, the issue with the `self.root.after` makes a lot of sense. The new behavior (overlays appear all at once, instead of litlte by little) is much better. Nice!

I'll merge this PR and release a new version. I'm in a period between jobs so things have been a little chaotic on my side, so I apologize for the delay once again. Please feel free to create PRs for any other features that you think would be welcome, and I'll make it a priority to respond to them in 2 or 3 days.

Cheers!
